src/eval_utils/run.ts

```diff
@@ -0,0 +1,725 @@
+                    inputs: request.inputs,
+                    messages: request.messages,
+                });
+
+            if (request.runId === undefined) {
+                request = {
+                    ...request,
+                    runId,
+                };
+            }
+            if (request.sourceDatapointId === undefined) {
+                request = {
+                    ...request,
+                    sourceDatapointId: sourceDatapointId,
+                };
+            }
+            if (client instanceof Flows) {
+                request = {
+                    ...request,
+                    traceStatus: "complete",
+                };
+            }
+
+            if ("flow" in request) {
+                if (!_.isEqual(state!.evaluatedVersion, request.flow)) {
+                    response = await originalLog(
+                        {
+                            ...request,
+                            // @ts-ignore Log under the version expected by evaluation, not
+                            // one determined by decorators. Otherwise the evaluation will stale
+                            flow: state?.evaluatedVersion,
+                            output: undefined,
+                            error: `The version of the evaluated Flow must match the version of the callable. Expected: ${JSON.stringify(state!.evaluatedVersion)}, got: ${JSON.stringify(request.flow)}`,
+                        },
+                        options,
+                    );
+                }
+            }
+
+            if ("prompt" in request) {
+                if (!_.isEqual(state!.evaluatedVersion, request.prompt)) {
+                    response = await originalLog({
+                        ...request,
+                        // @ts-ignore Log under the version expected by evaluation, not
+                        // one determined by decorators. Otherwise the evaluation will stale
+                        prompt: state?.evaluatedVersion,
+                        output: undefined,
+                        error: `The version of the evaluated Prompt must match the version of the callable. Expected: ${JSON.stringify(state!.evaluatedVersion)}, got: ${JSON.stringify(request.prompt)}`,
+                    });
+                }
+            }
+
+            if (response === undefined) {
+                // Version validation passed, make a normal request
+                response = await originalLog(request, options);
+            }
+
+            // @ts-ignore
+            uploadCallback(response.id);
+        } else {
+            // @ts-ignore
+            response = await originalLog(request, options);
+        }
+
+        return response;
+    };
+
+    client.log = _overloadedLog.bind(client);
+
+    return client;
+}
+
+export async function runEval(
+    client: HumanloopClient,
+    file: File,
+    dataset: Dataset,
+    name?: string,
+    evaluators: Evaluator[] = [],
+    workers: number = 8,
```

Comment:
"workers" here doesn't really make sense i think - does this actually spin off multiple workers? or is it just async concurrency?

does js have a standard name for this? (i do know this means inconsistency with python, but i think it's wrong enough that i'd push to change it here)

Reply:
p-map limits the number of promises running at any time to be `workers`.

I think it makes sense with an extra modification: we should cap the number of workers to not DOS our backend. Since the promises fire up simultaneously, we might have thousands of open HTTP connections and overwhelm the backend.

Added a 32 cap.

Reply:
my comment here was more about the specific terminology of "worker".

Reply:
looking at p-map, it looks like they use "concurrency", which i'd prefer here as it's more accurate, even if it's inconsistent with python. (seems reasonable as it's a language-specific runtime option)

Reply:
ack - I'll change to concurrency

Reply:
If "consistent with python" was a goal for this PR we should have closed it long ago 😄